src/server.c

```diff
@@ -744,7 +745,15 @@ int clientsCronResizeQueryBuffer(client *c) {
         /* There are two conditions to resize the query buffer: */
         if (idletime > 2) {
             /* 1) Query is idle for a long time. */
-            c->querybuf = sdsRemoveFreeSpace(c->querybuf, 1);
+            size_t remaining = sdslen(c->querybuf) - c->qb_pos;
+            if (!(c->flags & CLIENT_MASTER) && !remaining) {
```

Comment:
This is the only time we move a client back to the shared buffer, I don't quite follow why we only do it if the client is idle. Why wouldn't we move it back immediately?

Reply:
If we allocated a private buffer for the client, it means the client is using pipeline commands. In this case, it is reasonable to assume that the client will use pipeline commands again. To avoid excessive free-alloc cycles, the client will get back to a shared buffer only if it is idle. Basically, for a pipeline client, we keep the current behavior of using a private buffer until the client becomes idle.

Reply:
Can we document that logic here? I think it makes enough sense to me and don't have another suggestion.

Reply:
Added documentation.